Bugfixes + performance improvements

Samlibardi opened the pull request:
Resolvidas algumas exceptions causadas pela inicialização da database, cálculos de layout substituídos por soluções em CSS, resolvida issue #25 substituindo nested divs por canvas na grade de combinações.

Samlibardi committed:
Removed leftover properties that caused exceptions later on

Samlibardi committed:
Combination track performance changes
Scrolling now pure css, removed width calculation, scrolling to active combination now gets track children list instead of class lookup.

Samlibardi committed:
Now saves last color used on state
Last color used is saved so when page is reloaded the color counting continues properly

Samlibardi committed:
Search event won't fire until DB is loaded
This prevents undefined exception

Samlibardi committed:
combination HTML element now cached and created only once.

Samlibardi committed:
Combination boards are now canvas instead of div hierarchy. HUGE performance boost. Visually, no changes

Samlibardi committed:
Fixed timetable misalignment on printable pdf

Samlibardi committed:
PDF now opens in new tab
Instead of saving to "table.pdf", clicking the print button will now open a new tab with the PDF file, which can be printed directly from browser or saved as user wishes.

Samlibardi committed:
Hotfix for IE and Edge

pedropaulovc commented:
Olá Samuel! Obrigado novamente pelo pull request. Eu estou bem ocupado estes dias pra revisar o código. Prometo que vou dar uma olhada neste Sábado. Desculpe pela demora.

Samlibardi committed:
Added offline caching
Matrusp can now be used offline. All static files are cached until manifest file is changed. DB is updated when online and saved in local storage to be fetched when offline.

Samlibardi committed:
Changed html language to ptbr

Samlibardi committed:
Fixed word wrap in lecture explorer
Addresses #37. Layout wraps before title field gets too small. Added ellipsis when lecture is closed.

pedropaulovc commented on index.html:
```diff
@@ -1,5 +1,5 @@
-  <!DOCTYPE html>
-<html lang="en">
+<!DOCTYPE html>
+<html lang="pt_BR" manifest="matrusp.appcache">
```
De acordo com o MDN, application caching é uma feature _deprecada_: https://developer.mozilla.org/en-US/docs/Web/HTML/Using_the_application_cache 

Eu sugeriria que revertêssemos esta mudança. Uma alternativa seria adotar [Service Workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers) mas eu não sei o quão complicado seria. Outra alternativa, é manter as coisas como estão. A página do MatrUSP é simples o suficiente para funcionar apenas com caching HTTP.

Um outro adendo: ` +# Atualizar a data abaixo sempre que for feito deploy, para garantir que os arquivos cacheados serão atualizados.` Esse trecho me deixou preocupado. O processo de deploy atualmente é bem manual (basicamente tranferir arquivos via SCP). Para evitar erros humanos, o melhor é evitar que tenhamos que fazer mudanças manuais.

Samlibardi replied:
Pesquisei soluções para o caching inicialmente porque o caching HTTP estava ocasionando problemas no deploy, já que os arquivos estáticos não atualizam automaticamente e a única maneira de forçar um update é "enganar" o navegador colocando um parâmetro GET na tag HTML que chama os arquivos.. É um processo bastante inviável a não ser que seja automatizado.

Além disso, o caching HTTP não faz o cache de arquivos HTML, assim não é possível utilizar a página inteiramente off-line.

Em relação ao application caching e o service workers:
https://caniuse.com/#feat=serviceworkers tem apenas 64.69% + 9.14% (parcial) de suporte, não estando disponível no IE, e apenas nas versões de teste do Edge e do Safari, usando flags.
https://caniuse.com/#feat=offline-apps embora deprecado, o uso do cache manifest ainda é suportado por todos os maiores navegadores a partir do IE 10, o que resulta em 94.6% de suporte sem o uso de flags.
Além disso, o uso de Service Workers me pareceu uma solução bem mais complicada, enquanto o uso do application cache se resumiu à adição do arquivo manifest (além da parte da atualização programática do cache da database, pelas limitações do app cache).

Quanto a questão da modificação manual, acho que usar o manifest é a maneira mais simples e que requer menos manutenção: qualquer alteração no arquivo é suficiente para forçar o processo de atualização de todos os arquivos. Se for decidido por reverter essa mudança e usar apenas o cache HTTP, sugiro que ele seja desabilitado para que não haja problemas na distribuição dos arquivos estáticos do site.

pedropaulovc replied:
Dou um voto de confiança, vamos adotar a sua sugestão. :)

pedropaulovc commented on js/pdf.js:
```diff
@@ -107,11 +108,16 @@ function openpdf() {
-  combinations.style.display = "none";
+  combinations.style.visibility = "hidden";
+
+  if(!navigator.msSaveOrOpenBlob) var newTab = window.open('','_blank');
```
No Firefox e Chrome, o comportamento está um pouco estranho:

1. Monte a sua grade no ambiente de homologação (http://bcc.ime.usp.br/matrusp_homolog/)
2. Clique em imprimir
3. Uma nova aba em branco abre. Nada mais acontece.
4. Se você voltar para a aba do MatrUSP, a aba aberta passa a carregar o PDF.

Eu estava esperando que quando a nova aba abrir, o PDF já esteja gerado.

Samlibardi replied:
Isso acontece porque a geração do pdf só é concluída após a execução do html2canvas. Como ele funciona de maneira assíncrona, se a abertura de aba for feita no callback desse evento, ela é negada pelo bloqueador de pop-ups. O workaround é pedir que a abertura da aba seja feita durante o event listener do click e quando o pdf é finalizado, alterar o endereço da aba. Nos testes que fiz no chrome, não achei que o tempo de espera tenha sido muito grande mas me parece que o Firefox reduz significativamente a performance de javascript de abas em segundo plano (Se esperar alguns segundos, o PDF será carregado, embora ao voltar pra aba do MatrUSP o carregamento seja praticamente instantâneo). Nos dois casos, acho que a falta de aviso visual na nova aba é realmente um problema.

Com isso em vista, acredito que o melhor seja reverter a mudança e tentar uma solução mais bem polida.

Samlibardi replied:
Estou trabalhando em um rework do sistema de impressão.~~Será aplicada a classe "print" no body~~ _EDIT: é possível usar @media print_ e os elementos da página terão regras para obeceder isso, deixando a página formatada para impressão, e então será usado o comando window.print(), impressão nativa do navegador, que garante maior compatibilidade e consistência.

Esse trabalho está sendo feito em outro branch, quando finalizado criarei outro PR para ele. Vou reverter essa mudança por ora.

pedropaulovc commented on js/ui.js:
```diff
@@ -257,38 +257,15 @@ function UI() {
   this.createCombinationBoard = function(combination) {
     var combinationBoardTreeObj = {
-      tag: 'div',
+      tag: 'canvas',
```
A performance usando canvas ficou bem melhor, meus parabéns!

pedropaulovc replied:
A performance usando canvas ficou muito melhor. Parabéns!

pedropaulovc commented on js/ui.js:
```diff
@@ -257,38 +257,15 @@ function UI() {
 
   this.createCombinationBoard = function(combination) {
```
Não sei ao bem o que causou a regressão mas eu estou tendo problemas para carregar uma grade através de um identificador. Você pode confirmar se é relacionado com as  suas mudanças?

1. Monte uma grade no ambiente de homologação
2. Salve a grade com um identificador de sua preferência
3. Abra uma janela incognito no ambiente de homologação
4. Carregue o identificador

Resultado: Erro no console:

![image](https://user-images.githubusercontent.com/577970/35190311-ebbf16ee-fe12-11e7-89e5-9a45b8c9e924.png)

Samlibardi replied:
Me parece que isso está acontecendo por divergências entre os dados da disciplina salvos no state e os que estão no database. Me parece que esse método tester() tenta igualar os objetos do state aos do database, mas a chamada dele assume que eles tem o mesmo número de turmas, professores, etc.
Acredito que não seja relacionado às minhas mudanças, mas acho que posso tentar resolver o problema

Samlibardi replied:
Meu último commit deve resolver estes erros.

pedropaulovc replied:
Resolvido!

pedropaulovc reviewed (changes requested):
O PR está quase pronto para o merge. Apenas alguns detalhes a resolver. Por favor dê uma olhada. Valeu!

Samlibardi committed:
Fixed lecture explorer height and scrollbar, page height
Adresses #46: Lecture explorer scrollbar is now visible. Explorer no longer grows when lectures are opened. Page height is fixed so the scrollbar only appears when needed

Samlibardi committed:
Column background resizing now css instead of js

Samlibardi committed:
Revert "Hotfix for IE and Edge"

Samlibardi committed:
Revert "PDF now opens in new tab"

This reverts commit 44adf20ee9053664861329459c6ebfe86465eabe.

Samlibardi committed:
Fixed errors when loading state from server.

Samlibardi committed:
Improved db loading logic
Website loading time significantly improved. searchbox now respects DB async loading and indexing. However, IE 10 is probably unsupported

Samlibardi committed:
Pre-load db search queuing
Now it is possible to queue lecture search on DB to be executed when DB is loaded.

Samlibardi committed:
Updated html2canvas and jspdf
Solves schedule background rendering problems on print
